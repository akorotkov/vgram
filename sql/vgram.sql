CREATE EXTENSION vgram;

CREATE TABLE titles (s text);

\copy titles from 'data/titles.data'

SELECT qgram_stat(s, 2, 4, 0.05) AS vgrams FROM titles;
\gset

SELECT get_vgrams('indexing', 2, 4, :'vgrams');
SELECT get_vgrams('annotation', 2, 4, :'vgrams');
SELECT get_vgrams('i like it', 2, 4, :'vgrams');

CREATE INDEX titles_s_idx ON titles USING gin (s vgram_gin_ops (minq=-1, maxq=2));
CREATE INDEX titles_s_idx ON titles USING gin (s vgram_gin_ops (minq=2, maxq=11));
CREATE INDEX titles_s_idx ON titles USING gin (s vgram_gin_ops (minq=4, maxq=2));
CREATE INDEX titles_s_idx ON titles USING gin (s vgram_gin_ops (minq=2, maxq=4, vgrams=:'vgrams'));

SET enable_seqscan = ON;
SET enable_bitmapscan = OFF;

SELECT * FROM titles WHERE s ilike '%indexing%';
SELECT * FROM titles WHERE s ilike '%annotation%';

SET enable_seqscan = OFF;
SET enable_bitmapscan = ON;

EXPLAIN (COSTS OFF) SELECT * FROM titles WHERE s ilike '%indexing%';
EXPLAIN (COSTS OFF) SELECT * FROM titles WHERE s ilike '%annotation%';
SELECT * FROM titles WHERE s ilike '%indexing%';
SELECT * FROM titles WHERE s ilike '%annotation%';

DROP INDEX titles_s_idx;
CREATE INDEX titles_s_idx ON titles USING gin (s vgram_gin_ops (minq=3, maxq=3));

EXPLAIN (COSTS OFF) SELECT * FROM titles WHERE s ilike '%indexing%';
EXPLAIN (COSTS OFF) SELECT * FROM titles WHERE s ilike '%annotation%';
SELECT * FROM titles WHERE s ilike '%indexing%';
SELECT * FROM titles WHERE s ilike '%annotation%';

DROP INDEX titles_s_idx;
ALTER TABLE titles ALTER COLUMN s TYPE vgram_text USING s::vgram_text;
ANALYZE titles;
CREATE INDEX titles_s_idx ON titles USING gin (s vgram_gin_ops2 (minq=2, maxq=4, vgrams=:'vgrams'));

EXPLAIN (COSTS OFF) SELECT * FROM titles WHERE s ilike '%indexing%';
EXPLAIN (COSTS OFF) SELECT * FROM titles WHERE s ilike '%annotation%';
SELECT * FROM titles WHERE s ilike '%indexing%';
SELECT * FROM titles WHERE s ilike '%annotation%';
