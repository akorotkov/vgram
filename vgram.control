# vgram extension
comment = 'Variable-length grams'
default_version = '1.0'
module_pathname = '$libdir/vgram'
relocatable = true
