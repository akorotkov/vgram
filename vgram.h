/*-------------------------------------------------------------------------
 *
 * vgram.h
 *		Header for vgram module.
 *
 * Copyright (c) 2011-2017, Alexander Korotkov
 *
 * IDENTIFICATION
 *	  contrib/vgram/vgram.h
 *
 *-------------------------------------------------------------------------
 */
#ifndef _V_GRAM_H_
#define _V_GRAM_H_

#include "tsearch/ts_locale.h"

/*
 * V-gram parameters
 */
#define minQ						(2)
#define maxQ						(5)
#define isExtractable(c)			(isalpha((unsigned char) *(c)) || isdigit((unsigned char) *(c)))
#define VGRAM_LIMIT_RATIO			(0.005)
#define DEFAULT_CHARACTER_FREQUENCY	(0.001)
#define EMPTY_CHARACTER				('$')

/* strategy numbers */
#define LikeStrategyNumber			3
#define ILikeStrategyNumber			4


typedef void (*WordCallback) (const char *wordStart, const char *wordEnd, void *userData);
typedef void (*VGramCallBack) (char *vgram, void *userData);

typedef struct
{
	VGramCallBack	callback;
	void		   *userData;
} ExtractVGramsInfo;

extern void loadStats(void);
extern float4 estimateVGramSelectivilty(const char *vgram);
extern void extractMinimalVGramsWord(const char *wordStart, const char *wordEnd, void *userData);
extern void extractWords(const char *string, size_t len, WordCallback callback, void *userData);
extern void extractVGramsWord(const char *wordStart, const char *wordEnd, void *userData);
extern Datum *extractQueryLike(int32 *nentries, text *pattern);

#endif /* _V_GRAM_H_ */
