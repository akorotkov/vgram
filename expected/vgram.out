CREATE EXTENSION vgram;
CREATE TABLE titles (s text);
\copy titles from 'data/titles.data'
SELECT qgram_stat(s, 2, 4, 0.05) AS vgrams FROM titles;
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             vgrams                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 {$a,$a$,$an,$and,$ap,$app,$b,$ba,$bas,$c,$co,$com,$con,$d,$de,$di,$e,$f,$fo,$for,$g,$h,$ho,$hom,$i,$in,$in$,$int,$l,$m,$ma,$me,$mo,$mod,$n,$ne,$net,$o,$of,$of$,$on,$on$,$p,$pa,$pag,$pr,$pro,$r,$re,$s,$se,$st,$sy,$sys,$t,$te,$th,$the,$to,$to$,$tr,$u,$us,$v,$w,$wi,$wit,a$,ab,ac,act,ad,ag,age,age$,al,al$,am,an,an$,ana,and,and$,ap,app,ar,as,ase,ased,at,ate,ati,atio,ba,bas,base,bi,bl,c$,ca,cal,cat,cati,ce,ce$,ces,ch,ci,co,com,comp,con,ct,cti,ctio,cu,d$,da,de,del,di,du,e$,ea,ec,ect,ed,ed$,ee,el,el$,em,eme,en,ent,ent$,er,er$,era,eri,es,es$,ess,et,etw,ev,ex,f$,fe,fi,fo,for,for$,form,g$,ge,ge$,gen,gi,gn,gr,gra,h$,ha,he,he$,hi,ho,hom,home,ia,ic,ic$,ica,icat,id,ie,if,ig,il,im,in,in$,ine,ing,ing$,int,inte,io,ion,ion$,ions,ir,is,ist,it,ith,iti,ity,ity$,iv,ive,ive$,iz,k$,l$,la,le,le$,li,lin,ll,lo,lt,lu,ly,m$,ma,man,mat,mati,me,me$,men,ment,mi,mo,mod,mode,mp,ms,ms$,mu,mul,n$,na,nal,nc,nce,nce$,nd,nd$,ne,net,ng,ng$,ni,no,ns,ns$,nt,nt$,nte,o$,ob,oc,od,ode,odel,of,of$,og,ol,om,ome,ome$,omp,on,on$,ons,ons$,op,or,or$,ori,ork,orm,orma,os,ot,ou,pa,pag,page,pe,per,pl,po,pp,pr,pro,pt,qu,r$,ra,ral,rat,rc,re,re$,res,ri,rit,rk,rm,rma,ro,rs,rt,ry,s$,sc,se,sed,sed$,si,sin,sing,so,sp,ss,st,ste,stem,str,su,sy,sys,syst,t$,ta,tat,te,ted,ted$,tem,ter,th,th$,the,the$,ti,tic,tim,tin,ting,tio,tion,tiv,tive,to,to$,tor,tr,tra,tri,ts,ts$,tu,tur,tw,two,ty,ty$,ua,uc,ue,ul,ult,un,ur,ure,us,usi,usin,ut,va,ve,ve$,ver,vi,wi,wit,with,wo,wor,work,y$,ys,yst,yste}
(1 row)

\gset
SELECT get_vgrams('indexing', 2, 4, :'vgrams');
    get_vgrams    
------------------
 {ind,nde,dex,xi}
(1 row)

SELECT get_vgrams('annotation', 2, 4, :'vgrams');
    get_vgrams     
-------------------
 {nn,not,ota,tati}
(1 row)

SELECT get_vgrams('i like it', 2, 4, :'vgrams');
       get_vgrams        
-------------------------
 {$i$,$li,ik,ke,$it,it$}
(1 row)

CREATE INDEX titles_s_idx ON titles USING gin (s vgram_gin_ops (minq=-1, maxq=2));
ERROR:  value -1 out of bounds for option "minq"
DETAIL:  Valid values are between "1" and "10".
CREATE INDEX titles_s_idx ON titles USING gin (s vgram_gin_ops (minq=2, maxq=11));
ERROR:  value 11 out of bounds for option "maxq"
DETAIL:  Valid values are between "1" and "10".
CREATE INDEX titles_s_idx ON titles USING gin (s vgram_gin_ops (minq=4, maxq=2));
ERROR:  "minq" can't be greater than "maxq"
CREATE INDEX titles_s_idx ON titles USING gin (s vgram_gin_ops (minq=2, maxq=4, vgrams=:'vgrams'));
SET enable_seqscan = ON;
SET enable_bitmapscan = OFF;
SELECT * FROM titles WHERE s ilike '%indexing%';
                                              s                                               
----------------------------------------------------------------------------------------------
 Hybrid term indexing for different IR models.
 Word Spotting: A New Approach to Indexing Handwriting.
 Fast Discovery of Sequential Patterns through Memory Indexing and Database Partitioning.
 Efficient speech indexing and search for embedded devices using uniterms.
 Spatio-Temporal Indexing in Oracle: Issues and Challenges.
 Knowledge-Enhanced Latent Semantic Indexing.
 The EURATOM automatic indexing project.
 Spectral cross-correlation features for audio indexing of broadcast news and meetings.
 Crawling, Indexing, and Similarity Searching Images on the Web.
 Integrated Technologies for Indexing Spoken Language.
 Hybrid Transformation for Indexing and Searching Web Documents in the Cartographic Paradigm.
 Indexing Continual Range Queries with Covering Tiles for Fast Locating of Moving Objects.
 Indexing and Mining Time Series Data.
(13 rows)

SELECT * FROM titles WHERE s ilike '%annotation%';
                                                            s                                                            
-------------------------------------------------------------------------------------------------------------------------
 Information theory applied to the sparse gene ontology annotation network to predict novel gene function.
 An Image Annotation Guide Agent.
 Image Annotation System Using Visual and Textual Features.
 A Tool for Automatic Genome Annotation.
 Ubiquitous Interactive Video Editing Via Multimodal Annotations.
 Image annotation using multi-correlation probabilistic matrix factorization.
 Documenting Motion Sequences with a Personalized Annotation System.
 Semantic Annotation of a Natural Language Corpus for Knowledge Extraction.
 Semantic Video Annotation using Background Knowledge and Similarity-based Video Retrieval.
 An Efficient Time Annotation Technique in Abstract RTOS Simulations for Multiprocessor Task Migration.
 Using local discriminant topic to improve generative model based image annotation.
 Tag around: a 3D gesture game for image annotation.
 Evaluation of a large-scale biomedical data annotation initiative.
 Exploiting Term Co-occurrence for Enhancing Automated Image Annotation.
 Semi-automatic annotation system for home videos.
 Lazy Annotation for Program Testing and Verification.
 The value of adaptive link annotation in e-learning: a study of a portal-based approach.
 Effects of link annotations on search performance in layered and unlayered hierarchically organized information spaces.
 Models and Tools for Collaborative Annotation
(19 rows)

SET enable_seqscan = OFF;
SET enable_bitmapscan = ON;
EXPLAIN (COSTS OFF) SELECT * FROM titles WHERE s ilike '%indexing%';
                   QUERY PLAN                   
------------------------------------------------
 Bitmap Heap Scan on titles
   Recheck Cond: (s ~~* '%indexing%'::text)
   ->  Bitmap Index Scan on titles_s_idx
         Index Cond: (s ~~* '%indexing%'::text)
(4 rows)

EXPLAIN (COSTS OFF) SELECT * FROM titles WHERE s ilike '%annotation%';
                    QUERY PLAN                    
--------------------------------------------------
 Bitmap Heap Scan on titles
   Recheck Cond: (s ~~* '%annotation%'::text)
   ->  Bitmap Index Scan on titles_s_idx
         Index Cond: (s ~~* '%annotation%'::text)
(4 rows)

SELECT * FROM titles WHERE s ilike '%indexing%';
                                              s                                               
----------------------------------------------------------------------------------------------
 Hybrid term indexing for different IR models.
 Word Spotting: A New Approach to Indexing Handwriting.
 Fast Discovery of Sequential Patterns through Memory Indexing and Database Partitioning.
 Efficient speech indexing and search for embedded devices using uniterms.
 Spatio-Temporal Indexing in Oracle: Issues and Challenges.
 Knowledge-Enhanced Latent Semantic Indexing.
 The EURATOM automatic indexing project.
 Spectral cross-correlation features for audio indexing of broadcast news and meetings.
 Crawling, Indexing, and Similarity Searching Images on the Web.
 Integrated Technologies for Indexing Spoken Language.
 Hybrid Transformation for Indexing and Searching Web Documents in the Cartographic Paradigm.
 Indexing Continual Range Queries with Covering Tiles for Fast Locating of Moving Objects.
 Indexing and Mining Time Series Data.
(13 rows)

SELECT * FROM titles WHERE s ilike '%annotation%';
                                                            s                                                            
-------------------------------------------------------------------------------------------------------------------------
 Information theory applied to the sparse gene ontology annotation network to predict novel gene function.
 An Image Annotation Guide Agent.
 Image Annotation System Using Visual and Textual Features.
 A Tool for Automatic Genome Annotation.
 Ubiquitous Interactive Video Editing Via Multimodal Annotations.
 Image annotation using multi-correlation probabilistic matrix factorization.
 Documenting Motion Sequences with a Personalized Annotation System.
 Semantic Annotation of a Natural Language Corpus for Knowledge Extraction.
 Semantic Video Annotation using Background Knowledge and Similarity-based Video Retrieval.
 An Efficient Time Annotation Technique in Abstract RTOS Simulations for Multiprocessor Task Migration.
 Using local discriminant topic to improve generative model based image annotation.
 Tag around: a 3D gesture game for image annotation.
 Evaluation of a large-scale biomedical data annotation initiative.
 Exploiting Term Co-occurrence for Enhancing Automated Image Annotation.
 Semi-automatic annotation system for home videos.
 Lazy Annotation for Program Testing and Verification.
 The value of adaptive link annotation in e-learning: a study of a portal-based approach.
 Effects of link annotations on search performance in layered and unlayered hierarchically organized information spaces.
 Models and Tools for Collaborative Annotation
(19 rows)

DROP INDEX titles_s_idx;
CREATE INDEX titles_s_idx ON titles USING gin (s vgram_gin_ops (minq=3, maxq=3));
EXPLAIN (COSTS OFF) SELECT * FROM titles WHERE s ilike '%indexing%';
                   QUERY PLAN                   
------------------------------------------------
 Bitmap Heap Scan on titles
   Recheck Cond: (s ~~* '%indexing%'::text)
   ->  Bitmap Index Scan on titles_s_idx
         Index Cond: (s ~~* '%indexing%'::text)
(4 rows)

EXPLAIN (COSTS OFF) SELECT * FROM titles WHERE s ilike '%annotation%';
                    QUERY PLAN                    
--------------------------------------------------
 Bitmap Heap Scan on titles
   Recheck Cond: (s ~~* '%annotation%'::text)
   ->  Bitmap Index Scan on titles_s_idx
         Index Cond: (s ~~* '%annotation%'::text)
(4 rows)

SELECT * FROM titles WHERE s ilike '%indexing%';
                                              s                                               
----------------------------------------------------------------------------------------------
 Hybrid term indexing for different IR models.
 Word Spotting: A New Approach to Indexing Handwriting.
 Fast Discovery of Sequential Patterns through Memory Indexing and Database Partitioning.
 Efficient speech indexing and search for embedded devices using uniterms.
 Spatio-Temporal Indexing in Oracle: Issues and Challenges.
 Knowledge-Enhanced Latent Semantic Indexing.
 The EURATOM automatic indexing project.
 Spectral cross-correlation features for audio indexing of broadcast news and meetings.
 Crawling, Indexing, and Similarity Searching Images on the Web.
 Integrated Technologies for Indexing Spoken Language.
 Hybrid Transformation for Indexing and Searching Web Documents in the Cartographic Paradigm.
 Indexing Continual Range Queries with Covering Tiles for Fast Locating of Moving Objects.
 Indexing and Mining Time Series Data.
(13 rows)

SELECT * FROM titles WHERE s ilike '%annotation%';
                                                            s                                                            
-------------------------------------------------------------------------------------------------------------------------
 Information theory applied to the sparse gene ontology annotation network to predict novel gene function.
 An Image Annotation Guide Agent.
 Image Annotation System Using Visual and Textual Features.
 A Tool for Automatic Genome Annotation.
 Ubiquitous Interactive Video Editing Via Multimodal Annotations.
 Image annotation using multi-correlation probabilistic matrix factorization.
 Documenting Motion Sequences with a Personalized Annotation System.
 Semantic Annotation of a Natural Language Corpus for Knowledge Extraction.
 Semantic Video Annotation using Background Knowledge and Similarity-based Video Retrieval.
 An Efficient Time Annotation Technique in Abstract RTOS Simulations for Multiprocessor Task Migration.
 Using local discriminant topic to improve generative model based image annotation.
 Tag around: a 3D gesture game for image annotation.
 Evaluation of a large-scale biomedical data annotation initiative.
 Exploiting Term Co-occurrence for Enhancing Automated Image Annotation.
 Semi-automatic annotation system for home videos.
 Lazy Annotation for Program Testing and Verification.
 The value of adaptive link annotation in e-learning: a study of a portal-based approach.
 Effects of link annotations on search performance in layered and unlayered hierarchically organized information spaces.
 Models and Tools for Collaborative Annotation
(19 rows)

DROP INDEX titles_s_idx;
ALTER TABLE titles ALTER COLUMN s TYPE vgram_text USING s::vgram_text;
ANALYZE titles;
CREATE INDEX titles_s_idx ON titles USING gin (s vgram_gin_ops2 (minq=2, maxq=4, vgrams=:'vgrams'));
EXPLAIN (COSTS OFF) SELECT * FROM titles WHERE s ilike '%indexing%';
                   QUERY PLAN                   
------------------------------------------------
 Bitmap Heap Scan on titles
   Recheck Cond: (s ~~* '%indexing%'::text)
   ->  Bitmap Index Scan on titles_s_idx
         Index Cond: (s ~~* '%indexing%'::text)
(4 rows)

EXPLAIN (COSTS OFF) SELECT * FROM titles WHERE s ilike '%annotation%';
                    QUERY PLAN                    
--------------------------------------------------
 Bitmap Heap Scan on titles
   Recheck Cond: (s ~~* '%annotation%'::text)
   ->  Bitmap Index Scan on titles_s_idx
         Index Cond: (s ~~* '%annotation%'::text)
(4 rows)

SELECT * FROM titles WHERE s ilike '%indexing%';
                                              s                                               
----------------------------------------------------------------------------------------------
 Hybrid term indexing for different IR models.
 Word Spotting: A New Approach to Indexing Handwriting.
 Fast Discovery of Sequential Patterns through Memory Indexing and Database Partitioning.
 Efficient speech indexing and search for embedded devices using uniterms.
 Spatio-Temporal Indexing in Oracle: Issues and Challenges.
 Knowledge-Enhanced Latent Semantic Indexing.
 The EURATOM automatic indexing project.
 Spectral cross-correlation features for audio indexing of broadcast news and meetings.
 Crawling, Indexing, and Similarity Searching Images on the Web.
 Integrated Technologies for Indexing Spoken Language.
 Hybrid Transformation for Indexing and Searching Web Documents in the Cartographic Paradigm.
 Indexing Continual Range Queries with Covering Tiles for Fast Locating of Moving Objects.
 Indexing and Mining Time Series Data.
(13 rows)

SELECT * FROM titles WHERE s ilike '%annotation%';
                                                            s                                                            
-------------------------------------------------------------------------------------------------------------------------
 Information theory applied to the sparse gene ontology annotation network to predict novel gene function.
 An Image Annotation Guide Agent.
 Image Annotation System Using Visual and Textual Features.
 A Tool for Automatic Genome Annotation.
 Ubiquitous Interactive Video Editing Via Multimodal Annotations.
 Image annotation using multi-correlation probabilistic matrix factorization.
 Documenting Motion Sequences with a Personalized Annotation System.
 Semantic Annotation of a Natural Language Corpus for Knowledge Extraction.
 Semantic Video Annotation using Background Knowledge and Similarity-based Video Retrieval.
 An Efficient Time Annotation Technique in Abstract RTOS Simulations for Multiprocessor Task Migration.
 Using local discriminant topic to improve generative model based image annotation.
 Tag around: a 3D gesture game for image annotation.
 Evaluation of a large-scale biomedical data annotation initiative.
 Exploiting Term Co-occurrence for Enhancing Automated Image Annotation.
 Semi-automatic annotation system for home videos.
 Lazy Annotation for Program Testing and Verification.
 The value of adaptive link annotation in e-learning: a study of a portal-based approach.
 Effects of link annotations on search performance in layered and unlayered hierarchically organized information spaces.
 Models and Tools for Collaborative Annotation
(19 rows)

